pennylane/optimize/gradient_descent.py

```diff
@@ -79,14 +96,23 @@ def compute_grad(objective_fn, x, grad_fn=None):
                 If ``None``, the gradient function is computed automatically.
 
         Returns:
-            array: NumPy array containing the gradient :math:`\nabla f(x^{(t)})`
+            tuple: tuple with NumPy array containing the gradient :math:`\nabla f(x^{(t)})` and the
+                objective function output
         """
         if grad_fn is not None:
-            g = grad_fn(x)  # just call the supplied grad function
+            g = grad_fn  # just call the supplied grad function
+
+            try:
+                forward = g.forward
+            # if grad_fn is not an autograd gradient function, calculate the
+            # objective function output separately
+            except AttributeError:
+                forward = objective_fn(x)
         else:
             # default is autograd
-            g = autograd.grad(objective_fn)(x)  # pylint: disable=no-value-for-parameter
-        return g
+            g = get_gradient(objective_fn)  # pylint: disable=no-value-for-parameter
+            forward = g.forward
+        return g(x), forward
```

Comment:
Note that the forward value won't exist, or will refer to a previous step. It is only updated _after_ the gradient function is evaluated (e.g., `g(x)` will need to be evaluated prior to `g.forward` being accessed)

Reply:
Yes, that's true. With the simplified logic you proposed above though, I don't think this would ever be an issue here.